### ISSUE TITLE
Fix or remove _aligned_malloc()  and _aligned_free()

### ISSUE THREAD

**Issue by jsakkine** [kurasu/surge#171]
These do not seem to do anything useful. Can they be removed or not? Again purging unnecessary flags to make room for cmake.

**Comment by jsakkine** [comment]
PS. what about this after long description for a fix:

```
Fixes: #171 ("Fix or remove _aligned_malloc() and _aligned_free()")
```

**Comment by baconpaul** [comment]
So as I’d mentioned over on the draft of the version (but should put it on this ticket) why do you think they don’t do anything useful? To use SSE Intrinsics you need to be aligned; aligned_malloc guarantees that. I think we actually need these to have the SSE dsp work right?

**Comment by kurasu** [comment]
These are absolutely required on windows (or some other way to make sure these allocations are 16-byte aligned) or the process will terminate on SSE instructions.

The reason it's not generally needed on macOS is because the normal malloc already has 16-byte alignment.

There's a std::aligned_alloc now we could potentially use, but it requires c++17. Other option would be to use the alignas specifier on the classes/structs as needed and just use the normal new. But alas, you can't test that this works properly on mac.

**Comment by jsakkine** [comment]
Something like:

```
void* operator new[](size_t sz)
{
#if WINDOWS
   return _aligned_malloc(sz, 16);
#else 
   return malloc(sz);
#endif
}

void operator delete[](void* m)
{
#if WINDOWS
   return _aligned_free(sz);
#else 
   return free(sz);
#endif
}
```

**Comment by jsakkine** [comment]
Side-note: the above will require us to throw the exception (or do whatever we want) since we are not doing the default behavior.

**Comment by baconpaul** [comment]
Yeah but some news aren’t aligned. The code mixes aligned malloc and new.

Also how exhaustive is your windows test environment? Mine isn’t that great.

This seems like a high risk change to solve a problem we haven’t run into tbh. I would be super hesitant to merge a pr in this change space although @kurasu may!

**Comment by jsakkine** [comment]
Definitely will implement a PoC of this and test it on Windows. Not an urgent task to get upstream fast but absolutely worth of doing over time. If we want to do for example ARM/mobile port at some point you need to have the memory management better implemented.

**Comment by baconpaul** [comment]
Yup that’s right! But if you do that could you make prs from a branch rather than selecting commits from your master? I try to build all the prs so “3 commits on master I have to cherry pick” is way harder than “grab jsakkine/foobaz”

**Comment by jsakkine** [comment]
Yes, I can do it. And actually this kind of definitive change that I don't even want to do PR at instant but do git rebase for the branch for a while and just make it available.

**Comment by baconpaul** [comment]
Great. Thanks!

**Comment by kurasu** [comment]
surely this should do the trick?

class alignas(16) Oscillator

Then you can just use normal new for these classes and don't need to any global overrides of new & delete, which I find very dirty.

**Comment by baconpaul** [comment]
And then test like crazy on windows basically? Because linux and mac will do this naturally?

**Comment by kurasu** [comment]
If its not correct then it should crash at least half the time (per allocation), so it will be noticed pretty fast.

And I expect they at least tested the new operator when implementing the alignas specifier, so I think we should be fine.

**Comment by jsakkine** [comment]
@kurasu, interesting. I don't know but it is definitely worth of testing. I'll PoC that approach.

**Comment by jsakkine** [comment]
Example program:

```
#include <iostream>
#include <iomanip>

class alignas(256) Foo {
    unsigned long data;
};

int main()
{
    Foo *foo = new Foo();
    std::cout << std::internal << std::hex << std::setw(18) << std::setfill('0')
              << foo << std::endl;

    return 0;
}
```
If I compiled with `-faligned-new`:

```
$ g++ -faligned-new -o test test.cpp 

$ ./test 
0x0000557b8ff08f00
```
So yeah, looks really good.

Compiler support:
- GCC: since 7.0 [1]. Distributions currently ship with 8.x.
- Clang: since 4 0 [2]. Isn't this quite old already?
- Visual Studio: since 15.5. Not sure how recent this is.

[1] https://www.gnu.org/software/gcc/projects/cxx-status.html
[2] https://clang.llvm.org/cxx_status.html
[3] https://docs.microsoft.com/en-us/cpp/build/reference/zc-alignednew?view=vs-2017

**Comment by jsakkine** [comment]
Closing this one as this is now merged with #188.
